engine/classes/Elgg/Database/RelationshipsTable.php

```diff
@@ -20,12 +21,33 @@ class RelationshipsTable {
 	private $db;
 
 	/**
+	 * @var EntityTable
+	 */
+	private $entities;
+
+	/**
+	 * @var MetadataTable
+	 */
+	private $metadata;
+
+	/**
+	 * @var EventsService
+	 */
+	private $events;
+
+	/**
 	 * Constructor
 	 *
-	 * @param Database $db Elgg Database
+	 * @param Database      $db       Elgg Database
+	 * @param EntityTable   $entities Entity table 
+	 * @param MetadataTable $metadata Metadata table
+	 * @param EventsService $events   Events service
 	 */
-	public function __construct(Database $db) {
+	public function __construct(Database $db, EntityTable $entities, MetadataTable $metadata, EventsService $events) {
 		$this->db = $db;
+		$this->entities = $entities;
+		$this->metadata = $metadata;
+		$this->events = $events;
```

Comment:
Test are failing because this captures Elgg's global events instance, but we create a new one in `ElggRelationshipTest::setUp`. So in `\ElggRelationshipTest::testPreventAddRelationship` the event handler assignment and event trigger happen on different instances. Sane way to fix this is to replace this "clean events" pattern with use of https://github.com/Elgg/Elgg/pull/8949/files#diff-7f8c7b31ed7e60871431f36fa8c0ef90R264